File: examples/dekita/.claude/hooks/session_file_state_check.py
#!/usr/bin/env python3
# - 責務: セッション再開時にファイル状態（uncommitted changes）を検証
# - 重複なし: session-resume-warningは競合警告、こちらはファイル状態検証
# - 非ブロッキング: 情報表示のみ
"""
セッション再開時ファイル状態検証フック（SessionStart）

Issue #2468: セッション再開時にサマリーと実際のファイル状態が乖離していることがある。
- サマリーには「編集完了」と記載されているが、実際は未コミット
- サマリーを信頼して次のステップに進もうとすると問題が発生

このフックは:
1. `git status` でuncommitted changesを確認
2. セッション再開時（resume/compact）に未コミット変更があれば警告を表示
3. 直前のコミット内容を表示して整合性確認を促す
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))
from lib.execution import log_hook_execution
from lib.session import parse_hook_input

HOOK_NAME = "session-file-state-check"


def get_git_status() -> dict[str, list[str]]:
    """Get uncommitted changes from git status.

    Returns:
        Dict with 'staged', 'unstaged', and 'untracked' file lists.
    """
    result = {"staged": [], "unstaged": [], "untracked": []}

    try:
        status_result = subprocess.run(
            ["git", "status", "--porcelain", "-z"],
            capture_output=True,
            text=True,
            check=False,
            timeout=5,
        )
        if status_result.returncode != 0:
            return result

        # -z option: NUL-separated entries, handles special characters in filenames
        # Format: XY filename\0 (or XY oldname\0newname\0 for renames)
        entries = status_result.stdout.split("\0")
        i = 0
        while i < len(entries):
            entry = entries[i]
            if not entry:
                i += 1
                continue

            # Porcelain format: XY filename
            # X = index status, Y = work tree status
            index_status = entry[0] if len(entry) > 0 else " "
            worktree_status = entry[1] if len(entry) > 1 else " "
            filename = entry[3:] if len(entry) > 3 else ""

            if index_status == "?":
                result["untracked"].append(filename)
            elif index_status != " ":
                result["staged"].append(filename)
            if worktree_status not in (" ", "?"):
                result["unstaged"].append(filename)

            # Handle renames (R) and copies (C) which have a second filename
            if index_status in ("R", "C"):
                i += 1  # Skip the next entry (old filename)

            i += 1

        return result
    except Exception:
        return result


def get_last_commit_info() -> str | None:
    """Get the last commit message and affected files.

    Returns:
        Formatted string with commit info, or None if unavailable.
    """
    try:
        # Get last commit hash, message, and time
        log_result = subprocess.run(
            [
                "git",
                "log",
                "-1",
                "--format=%h %s (%ar)",
            ],
            capture_output=True,
            text=True,
            check=False,
            timeout=5,
        )
        if log_result.returncode != 0 or not log_result.stdout.strip():
            return None

        commit_info = log_result.stdout.strip()

        # Get files changed in last commit (-z for NUL-separated output)
        files_result = subprocess.run(
            ["git", "diff-tree", "--no-commit-id", "--name-only", "-r", "-z", "HEAD"],
            capture_output=True,
            text=True,
            check=False,
            timeout=5,
        )
        files = []
        if files_result.returncode == 0:
            files = [f for f in files_result.stdout.split("\0") if f][:5]

        result = f"  {commit_info}"
        if files:
            files_str = ", ".join(files)
            if len(files) >= 5:
                files_str += ", ..."
            result += f"\n  変更ファイル: {files_str}"

        return result
    except Exception:
        return None


def format_file_state_warning(
    status: dict[str, list[str]],
    last_commit: str | None,
) -> str:
    """Format the file state warning message.

    Args:
        status: Dict with staged, unstaged, untracked file lists.
        last_commit: Last commit info string.
    """
    message_parts = [
        "⚠️ **ファイル状態の確認が必要です**\n",
        "セッション再開時に未コミットの変更が検出されました。",
        "**サマリーとファイル状態が乖離している可能性があります**。\n",
    ]

    # Show uncommitted changes
    if status["staged"]:
        message_parts.append(f"**ステージ済み** ({len(status['staged'])}件):")
        for f in status["staged"][:5]:
            message_parts.append(f"  - {f}")
        if len(status["staged"]) > 5:
            message_parts.append(f"  ... 他 {len(status['staged']) - 5}件")
        message_parts.append("")

    if status["unstaged"]:
        message_parts.append(f"**未ステージ変更** ({len(status['unstaged'])}件):")
        for f in status["unstaged"][:5]:
            message_parts.append(f"  - {f}")
        if len(status["unstaged"]) > 5:
            message_parts.append(f"  ... 他 {len(status['unstaged']) - 5}件")
        message_parts.append("")

    if status["untracked"]:
        message_parts.append(f"**未追跡ファイル** ({len(status['untracked'])}件):")
        for f in status["untracked"][:3]:
            message_parts.append(f"  - {f}")
        if len(status["untracked"]) > 3:
            message_parts.append(f"  ... 他 {len(status['untracked']) - 3}件")
        message_parts.append("")

    # Show last commit for comparison
    if last_commit:
        message_parts.append("**直前のコミット**:")
        message_parts.append(last_commit)
        message_parts.append("")

    # Add guidance
    message_parts.extend(
        [
            "📋 **確認事項**:",
            "- サマリーの「完了」項目が実際にコミット済みか確認",
            "- 未コミット変更がサマリーの作業内容と一致するか確認",
            "- 不整合がある場合、`git status` と `git diff` で詳細確認",
        ]
    )

    return "\n".join(message_parts)


def main():
    """SessionStart hookのエントリーポイント"""
    result = {"continue": True}

    try:
        hook_input = parse_hook_input()
        source = hook_input.get("source", "")

        # Only check on session resume (resume or compact)
        if source not in ("resume", "compact"):
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"Not a resume session (source={source})",
            )
            print(json.dumps(result))
            return

        # Get git status
        status = get_git_status()
        has_changes = any((status["staged"], status["unstaged"], status["untracked"]))

        # If no changes at all, nothing to warn about
        if not has_changes:
            log_hook_execution(
                HOOK_NAME,
                "approve",
                "Working tree is clean",
            )
            print(json.dumps(result))
            return

        # Get last commit info for context
        last_commit = get_last_commit_info()

        # Format and display warning
        result["message"] = format_file_state_warning(status, last_commit)

        log_hook_execution(
            HOOK_NAME,
            "approve",
            f"Uncommitted changes detected (staged={len(status['staged'])}, "
            f"unstaged={len(status['unstaged'])}, untracked={len(status['untracked'])})",
            details={
                "source": source,
                "staged_count": len(status["staged"]),
                "unstaged_count": len(status["unstaged"]),
                "untracked_count": len(status["untracked"]),
            },
        )

    except Exception as e:
        log_hook_execution(
            HOOK_NAME,
            "approve",
            f"Error: {e}",
        )

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/session_file_state_check.ts
#!/usr/bin/env bun
/**
 * セッション再開時ファイル状態検証フック（SessionStart）
 *
 * Why:
 *   セッション再開時にサマリーと実際のファイル状態が乖離していることがある。
 *   サマリーには「編集完了」と記載されているが、実際は未コミットの場合、
 *   サマリーを信頼して次のステップに進むと問題が発生する。
 *
 * What:
 *   - `git status` でuncommitted changesを確認
 *   - セッション再開時（resume/compact）に未コミット変更があれば警告を表示
 *   - 直前のコミット内容を表示して整合性確認を促す
 *
 * State:
 *   - reads: none (git commandのみ)
 *
 * Remarks:
 *   - 非ブロック型（情報表示のみ）
 *   - 責務: セッション再開時にファイル状態を検証
 *   - session-resume-warningは競合警告、こちらはファイル状態検証
 *   - Python版: session_file_state_check.py
 *
 * Changelog:
 *   - silenvx/dekita#2468: フック追加
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { TIMEOUT_LIGHT } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "session-file-state-check";

export interface GitStatus {
  staged: string[];
  unstaged: string[];
  untracked: string[];
}

/**
 * Get uncommitted changes from git status.
 */
function getGitStatus(): GitStatus {
  const result: GitStatus = { staged: [], unstaged: [], untracked: [] };

  try {
    const statusResult = execSync("git status --porcelain -z", {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    // -z option: NUL-separated entries, handles special characters in filenames
    // Format: XY filename\0 (or XY oldname\0newname\0 for renames)
    const entries = statusResult.split("\0");
    let i = 0;

    while (i < entries.length) {
      const entry = entries[i];
      if (!entry) {
        i++;
        continue;
      }

      // Porcelain format: XY filename
      // X = index status, Y = work tree status
      const indexStatus = entry.length > 0 ? entry[0] : " ";
      const worktreeStatus = entry.length > 1 ? entry[1] : " ";
      const filename = entry.length > 3 ? entry.slice(3) : "";

      if (indexStatus === "?") {
        result.untracked.push(filename);
      } else if (indexStatus !== " ") {
        result.staged.push(filename);
      }

      if (worktreeStatus !== " " && worktreeStatus !== "?") {
        result.unstaged.push(filename);
      }

      // Handle renames (R) and copies (C) which have a second filename
      if (indexStatus === "R" || indexStatus === "C") {
        i++; // Skip the next entry (old filename)
      }

      i++;
    }
  } catch {
    // Return empty result on error
  }

  return result;
}

/**
 * Get the last commit message and affected files.
 */
function getLastCommitInfo(): string | null {
  try {
    // Get last commit hash, message, and time
    const logResult = execSync('git log -1 --format="%h %s (%ar)"', {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    if (!logResult) {
      return null;
    }

    // Get files changed in last commit (-z for NUL-separated output)
    let files: string[] = [];
    try {
      const filesResult = execSync("git diff-tree --no-commit-id --name-only -r -z HEAD", {
        encoding: "utf-8",
        timeout: TIMEOUT_LIGHT * 1000,
        stdio: ["pipe", "pipe", "pipe"],
      });
      files = filesResult
        .split("\0")
        .filter((f) => f)
        .slice(0, 5);
    } catch {
      // Ignore file listing errors
    }

    let result = `  ${logResult}`;
    if (files.length > 0) {
      let filesStr = files.join(", ");
      if (files.length >= 5) {
        filesStr += ", ...";
      }
      result += `\n  変更ファイル: ${filesStr}`;
    }

    return result;
  } catch {
    return null;
  }
}

/**
 * Format the file state warning message.
 */
export function formatFileStateWarning(status: GitStatus, lastCommit: string | null): string {
  const parts: string[] = [
    "⚠️ **ファイル状態の確認が必要です**\n",
    "セッション再開時に未コミットの変更が検出されました。",
    "**サマリーとファイル状態が乖離している可能性があります**。\n",
  ];

  // Show uncommitted changes
  if (status.staged.length > 0) {
    parts.push(`**ステージ済み** (${status.staged.length}件):`);
    for (const f of status.staged.slice(0, 5)) {
      parts.push(`  - ${f}`);
    }
    if (status.staged.length > 5) {
      parts.push(`  ... 他 ${status.staged.length - 5}件`);
    }
    parts.push("");
  }

  if (status.unstaged.length > 0) {
    parts.push(`**未ステージ変更** (${status.unstaged.length}件):`);
    for (const f of status.unstaged.slice(0, 5)) {
      parts.push(`  - ${f}`);
    }
    if (status.unstaged.length > 5) {
      parts.push(`  ... 他 ${status.unstaged.length - 5}件`);
    }
    parts.push("");
  }

  if (status.untracked.length > 0) {
    parts.push(`**未追跡ファイル** (${status.untracked.length}件):`);
    for (const f of status.untracked.slice(0, 3)) {
      parts.push(`  - ${f}`);
    }
    if (status.untracked.length > 3) {
      parts.push(`  ... 他 ${status.untracked.length - 3}件`);
    }
    parts.push("");
  }

  // Show last commit for comparison
  if (lastCommit) {
    parts.push("**直前のコミット**:");
    parts.push(lastCommit);
    parts.push("");
  }

  // Add guidance
  parts.push(
    "📋 **確認事項**:",
    "- サマリーの「完了」項目が実際にコミット済みか確認",
    "- 未コミット変更がサマリーの作業内容と一致するか確認",
    "- 不整合がある場合、`git status` と `git diff` で詳細確認",
  );

  return parts.join("\n");
}

async function main(): Promise<void> {
  const result: { continue: boolean; message?: string } = { continue: true };
  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    sessionId = inputData.session_id;
    const source = (inputData.source as string) || "";

    // Only check on session resume (resume or compact)
    if (source !== "resume" && source !== "compact") {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Not a resume session (source=${source})`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Get git status
    const status = getGitStatus();
    const hasChanges =
      status.staged.length > 0 || status.unstaged.length > 0 || status.untracked.length > 0;

    // If no changes at all, nothing to warn about
    if (!hasChanges) {
      await logHookExecution(HOOK_NAME, "approve", "Working tree is clean", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // Get last commit info for context
    const lastCommit = getLastCommitInfo();

    // Format and display warning
    result.message = formatFileStateWarning(status, lastCommit);

    await logHookExecution(
      HOOK_NAME,
      "approve",
      `Uncommitted changes detected (staged=${status.staged.length}, unstaged=${status.unstaged.length}, untracked=${status.untracked.length})`,
      {
        source,
        staged_count: status.staged.length,
        unstaged_count: status.unstaged.length,
        untracked_count: status.untracked.length,
      },
      { sessionId },
    );
  } catch (error) {
    await logHookExecution(HOOK_NAME, "approve", `Error: ${formatError(error)}`, undefined, {
      sessionId,
    });
  }

  console.log(JSON.stringify(result));
}

// Only run main when executed directly, not when imported
if (import.meta.main) {
  main();
}
